### PR TITLE
[ROCm] Route DSA paged top-k through aiter's one-block kernel

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Motivation

At decode, `topk_transform_decode_kernel` launches one workgroup per query row
(`sgl-kernel/csrc/.../topk.hip`: `grid = dim3{B}`). GLM-5.2 DSA decode runs at
batch 4 with 4 draft tokens, so that is 16 workgroups on a 256-CU MI355X — the
kernel is parallelism-starved rather than bandwidth- or compute-bound.

aiter's `top_k_per_row_decode` does the same selection with a wider radix
(11-12 bits vs 8, so 3 passes instead of 4). Measured on MI355X at the shapes
this path sees: **3.3x** at the p50 context and **3.9x** at p90.

## Modification

`_aiter_paged_topk_transform` routes the PAGED top-k transform to
`aiter.top_k_per_row_decode` on ROCm, and returns `None` — falling back to
sgl_kernel unchanged — for any shape the mapping does not cover:

- non-fp32 logits, or ragged rows (`row_starts is not None`)
- a page table whose row count does not match the logits row count
- `num_rows != batch * next_n`, where `next_n` is derived as `sum(extend) // batch`,
  the same derivation `dsa_indexer.py` uses (the two metadata builders disagree on
  the list layout but agree on the sum)

Selection is clamped to the page-table width before the gather, and the
clamp/gather/mask glue is fused into one Triton kernel instead of five launches.

`top_k_per_row_decode` is used rather than either multi-block entry on purpose:
it is one workgroup per row, so there is no cross-block synchronisation. aiter
[#3773](https://github.com/ROCm/aiter/pull/3773) made that entry unconditionally
one-block; the aiter commit pinned in `docker/rocm.Dockerfile` includes it.

## Accuracy / Performance

End-to-end, GLM-5.2-MXFP4 TP4/EP4 + MTP, agentic-coding trace replay, conc=4,
~1200 s per arm on MI355X:

| | ITL p50 | ITL mean | tput/GPU |
|---|---|---|---|
| baseline | 9.30 ms | 10.03 ms | 4158.5 |
| this PR | 8.88 ms | 9.88 ms | 4390.5 |
| | **-4.5%** | -1.5% | **+5.6%** |

Caveat on the end-to-end numbers: the two arms are separate runs, not
interleaved. On this part the shared power budget drifts a fixed kernel by up to
13% between runs, so the +5.6% sits close to that floor. The kernel-level 3.3-3.9x
is the measurement this change rests on.

## Checklist

- [x] Unit tests: `test/registered/unit/layers/test_dsa_topk_aiter_oneblock.py`
      (10 tests, CPU CI, aiter mocked) cover the fallback conditions and the
      `next_n` derivation under both metadata layouts.
- [x] Format: `black`, `ruff --select=F401,F821,UP037`.
- [x] No change to the imports or to any non-ROCm path.

---
### CI States

Latest PR Test (Base): :x: [Run #34805011141](https://github.com/sgl-project/sglang/actions/runs/34805011141)
Latest PR Test (Extra): :x: [Run #34805010860](https://github.com/sgl-project/sglang/actions/runs/34805010860)
Latest PR Test (AMD ROCm 10): :x: [Run #34805011160](https://github.com/sgl-project/sglang/actions/runs/34805011160)

### CODE REVIEW

**Comment by JohnQinAMD** [comment]
Closing this: for the model it was written and measured against, the code path it optimizes no longer runs.

## Why

This PR inserts an aiter fast path inside the `is_sgl_kernel()` PAGED branch of `DSATopKBackend.topk_transform`, i.e. the v1 path built on `fast_topk_transform_fused`. That branch is only reached when the top-k v2 dispatch above it declines, which is governed by `SGLANG_OPT_USE_TOPK_V2`.

When I opened this on 2026-08-27, `scripts/ci/slurm/launch_mi355x.sh` passed `SGLANG_OPT_USE_TOPK_V2=false`, so GLM-5.x on ROCm did run v1 and the fast path was live.

#36851 ("[AMD] Enable topk v2 GLM ROCm", merged 2026-09-01) changed that. `_handle_model_specific_adjustments` now does:

```python
if get_platform().is_hip:
    if is_deepseek_dsa(hf_config):
        envs.SGLANG_OPT_USE_TOPK_V2.set(False)
    if model_arch == "GlmMoeDsaForCausalLM":
        envs.SGLANG_OPT_USE_TOPK_V2.set(True)
```

`EnvBool.set()` writes `os.environ` directly, so for `GlmMoeDsaForCausalLM` this overrides the `=false` the launch script passes. GLM-5.x on ROCm now takes v2 unconditionally, and the branch this PR touches is dead for that family.

#36851 also reports a better result than this PR could have delivered — median ITL down 13.6-13.9% at concurrency >= 8 end to end — because it replaces v1 rather than making v1's kernel faster. The 3.3-3.9x here was a kernel-level comparison against `topk_transform_decode_kernel`, which is no longer what runs.

## What remains, for anyone who picks this up

The path is not dead for every model. On ROCm the blanket disable still applies to DSA models other than `GlmMoeDsaForCausalLM` (DeepSeek-V3.2, Glm5Next), so those still run v1 and would still reach this dispatch. Two things would need doing before the patch is usable there:

1. Re-measure. The 3.3-3.9x was taken at GLM-5.2 decode shapes on MI355X; nothing here establishes it at other models' shapes.

2. Fix the ragged guard. `next_n = sum(extend) // batch` with the check `num_rows == batch * next_n` does not reject ragged expansions, contrary to its comment: `extend=[5,3]` gives `batch=2, num_rows=8, next_n=4`, so `batch * next_n == num_rows` passes, and aiter's `rowEnds[batch_id / next_n]` row mapping then attributes rows to the wrong requests. `extend=[6,2]` and `extend=[4,2,3]` pass the same way. A uniformity check (`len(set(extend)) == 1`) is what the comment intends.

One part did hold up, and is worth recording: the per-row window semantics match exactly. aiter computes `row_len = rowEnds[batch_id / next_n] - next_n + (batch_id % next_n) + 1`, and `seqlens_expand_kernel` computes `start = kv_len - qo_len + 1` with row `j` at `start + j`. Substituting `batch_id = b * next_n + j` makes them identical, and the `-1` padding convention agrees on both sides.
